File: 第9章 数学游戏/9.8 兰顿蚂蚁/兰顿蚂蚁.cpp
// 蚂蚁爬行
void ant_run(vector<int> ant_map, int step, int cols, int rows)
{
	while (true) {
		float row = (pen.getY() + 180) / step;
		float col = (pen.getX() + 240) / step;
		if (row > rows || col > cols || row < 0 || col < 0)
			break;
		int index = (row - 1) * cols + col;
		if (0 == ant_map[index]) {
			ant_map[index] = 1;
			pen.color(_black);
			pen.lt(90);
		}
		else {
			ant_map[index] = 0;
			pen.color(_white);
			pen.rt(90);
		}
		pen.rr(step, step);
		pen.fd(step);
	}
}

// 兰顿蚂蚁游戏
int main()
{
	int step = 3;          // 格子大小
	int cols = 480 / step; // 地图的列数
	int rows = 360 / step; // 地图的行数
	// 初始化蚂蚁爬行的地图
	vector<int> ant_map(rows * cols + 1, 0);
	// 初始化画笔
	pen.speed(10);
	pen.up();
	pen.angle(0);
	pen.move(0, 0);
	pen.down();
	// 蚂蚁爬行
	ant_run(ant_map, step, cols, rows);
    return 0;
}

File: 第9章 数学游戏/9.8 兰顿蚂蚁/兰顿蚂蚁.py
from turtle_plus import *
# 全局变量
STEP = 3
COLS = 480 // STEP
ROWS = 360 // STEP
UP, RIGHT, DOWN, LEFT = 1, 2, 3, 4
    
class LangtonAnt:
    '''兰顿蚂蚁'''
    def __init__(self):
        '''初始化'''
        self.orientation = UP    # 蚂蚁向上移动
        self.ant_col = COLS // 2 # 蚂蚁放在中心
        self.ant_row = ROWS // 2
        self.ant_map = []        # 蚂蚁的地图
        for i in range(ROWS):
            self.ant_map.append([0] * COLS)

    def turn(self, direction):
        '''蚂蚁转弯'''
        if direction == 'right':
            self.orientation += 1
            
        if self.orientation > 4:
            self.orientation = 1
            
        if direction == 'left':
            self.orientation -= 1
            
        if self.orientation < 1:
            self.orientation = 4

    def forward(self):
        '''蚂蚁前进'''
        if self.orientation == UP:
            self.ant_row += 1
            
        if self.orientation == DOWN:
            self.ant_row -= 1
            
        if self.orientation == RIGHT:
            self.ant_col += 1
            
        if self.orientation == LEFT:
            self.ant_col -= 1
            
    def draw_dot(self):
        '''画点'''
        x = self.ant_col * STEP - 240 + STEP / 2
        y = self.ant_row * STEP - 180 + STEP / 2
        up()
        goto(x, y)
        down()
        dot(STEP)
    
    def run(self):
        '''蚂蚁爬行'''
        while True:
            if not (0 <= self.ant_row < ROWS and 0 <= self.ant_col < COLS):
                break
            
            if 0 == self.ant_map[self.ant_row][self.ant_col]:
                self.ant_map[self.ant_row][self.ant_col] = 1
                pencolor('black')
                self.turn('left')
            else:
                self.ant_map[self.ant_row][self.ant_col] = 0
                pencolor('white')
                self.turn('right')

            self.draw_dot()
            self.forward()
        done()

if __name__ == '__main__':
    '''初始化'''
    up()
    goto(0, 0)
    '''调用兰顿蚂蚁'''
    ant = LangtonAnt()
    ant.run()
